src/material/select/select.ts

```diff
@@ -317,12 +323,14 @@ export abstract class _MatSelectBase<C> extends _MatSelectMixinBase implements A
 
   /** Whether the component is required. */
   @Input()
-  get required(): boolean { return this._required; }
+  get required(): boolean {
+    return this._required ?? this.ngControl?.control?.hasValidator(Validators.required) ?? false;
```

Comment:
Should this also have a test for explicitly setting `required` overriding the form control required state?

Reply:
The forms module will log a deprecation warning if `required` is used on an element with `formControl`.